File: backend/internal/handler/websocket.go
package handler

import (
	"chatapp/internal/model"
	"chatapp/internal/service"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
)

var upgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool {
		return true
	},
}

type WSClient struct {
	conn   *websocket.Conn
	send   chan []byte
	userID int
}

type WSHub struct {
	clients    map[int]*WSClient
	broadcast  chan []byte
	register   chan *WSClient
	unregister chan *WSClient
}

func NewWSHub() *WSHub {
	return &WSHub{
		clients:    make(map[int]*WSClient),
		broadcast:  make(chan []byte),
		register:   make(chan *WSClient),
		unregister: make(chan *WSClient),
	}
}

func (h *WSHub) Run() {
	for {
		select {
		case client := <-h.register:
			h.clients[client.userID] = client
		case client := <-h.unregister:
			if _, ok := h.clients[client.userID]; ok {
				delete(h.clients, client.userID)
				close(client.send)
			}
		case message := <-h.broadcast:
			for _, client := range h.clients {
				select {
				case client.send <- message:
				default:
					close(client.send)
					delete(h.clients, client.userID)
				}
			}
		}
	}
}

type WebSocketHandler struct {
	hub            *WSHub
	chatService    *service.ChatService
	messageService *service.MessageService
}

func NewWebSocketHandler(chatService *service.ChatService, messageService *service.MessageService, userRepo interface{}) *WebSocketHandler {
	hub := NewWSHub()
	go hub.Run() // Start hub in background
	return &WebSocketHandler{
		hub:            hub,
		chatService:    chatService,
		messageService: messageService,
	}
}

func (h *WebSocketHandler) ServeWebSocket(c *gin.Context) {
	conn, err := upgrader.Upgrade(c.Writer, c.Request, nil)
	if err != nil {
		return
	}

	// Get user ID from query
	userIDStr := c.Query("userId")
	userID, err := strconv.Atoi(userIDStr)
	if err != nil || userID == 0 {
		conn.Close()
		return
	}

	client := &WSClient{
		conn:   conn,
		send:   make(chan []byte, 256),
		userID: userID,
	}

	h.hub.register <- client
	go h.writePump(client)
	go h.readPump(client)
}

func (h *WebSocketHandler) writePump(client *WSClient) {
	defer client.conn.Close()
	for {
		message, ok := <-client.send
		if !ok {
			client.conn.WriteMessage(websocket.CloseMessage, []byte{})
			return
		}
		client.conn.WriteMessage(websocket.TextMessage, message)
	}
}

func (h *WebSocketHandler) readPump(client *WSClient) {
	defer func() {
		h.hub.unregister <- client
		client.conn.Close()
	}()

	for {
		_, message, err := client.conn.ReadMessage()
		if err != nil {
			break
		}

		var msg struct {
			Type    string `json:"type"`
			ChatID  int    `json:"chat_id"`
			Content string `json:"content"`
		}
		if err := json.Unmarshal(message, &msg); err != nil {
			continue
		}

		if msg.Type == "message" {
			m := &model.Message{
				ChatID:  msg.ChatID,
				UserID:  client.userID,
				Content: msg.Content,
			}
			h.messageService.SendMessage(m)
			m.CreatedAt = time.Now()

			// Get username
			user, _ := h.chatService.GetUserByID(client.userID)
			if user != nil {
				m.Username = user.Username
			}

			data, _ := json.Marshal(m)
			h.hub.broadcast <- data
		}
	}
}

File: backend/handlers/websocket.py
import json

from fastapi import WebSocket

from models.models import Message
from services.chat_service import ChatService
from services.message_service import MessageService


class WSClient:
    def __init__(self, websocket: WebSocket, user_id: int):
        self.websocket = websocket
        self.user_id = user_id
        self.send_queue: list[bytes] = []


class WSHub:
    def __init__(self):
        self.clients: dict[int, WSClient] = {}
        self.broadcast_queue: list[tuple[int, bytes]] = []

    def register(self, client: WSClient):
        self.clients[client.user_id] = client

    def unregister(self, user_id: int):
        if user_id in self.clients:
            del self.clients[user_id]

    def add_broadcast(self, chat_id: int, data: bytes):
        self.broadcast_queue.append((chat_id, data))

    async def process_broadcasts(self):
        """Process queued broadcasts and send to relevant clients."""
        while self.broadcast_queue:
            chat_id, data = self.broadcast_queue.pop(0)
            # Broadcast to all connected clients (frontend will filter by chat_id)
            for client in list(self.clients.values()):
                try:
                    await client.websocket.send_bytes(data)
                except Exception:
                    self.unregister(client.user_id)


class WebSocketHandler:
    def __init__(self, chat_service: ChatService, message_service: MessageService, hub: WSHub):
        self.chat_service = chat_service
        self.message_service = message_service
        self.hub = hub

    async def connect(self, websocket: WebSocket, user_id: int):
        await websocket.accept()
        client = WSClient(websocket, user_id)
        self.hub.register(client)

    async def disconnect(self, user_id: int):
        self.hub.unregister(user_id)

    async def handle_message(self, user_id: int, chat_id: int, content: str):
        msg = Message(
            id=0, chat_id=chat_id, user_id=user_id,
            content=content, file_url=None,
            created_at=None, username=None
        )

        await self.message_service.send_message(msg)

        # Get username
        user = await self.chat_service.get_user_by_id(user_id)
        if user:
            msg.username = user.username

        # Serialize and broadcast
        data = json.dumps({
            "id": msg.id,
            "chat_id": msg.chat_id,
            "user_id": msg.user_id,
            "content": msg.content,
            "file_url": msg.file_url,
            "created_at": msg.created_at.isoformat() if msg.created_at else None,
            "username": msg.username
        })
        self.hub.add_broadcast(chat_id, data.encode("utf-8"))
